### PR TITLE
Making the verification and password reset token unqiue

### PR DESCRIPTION
**Pull request by mamazu**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.13                   |
| Bug fix?        | performance fix                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no |
| Related tickets | -                      |
| License         | MIT                                                          |

## The problem
Currently we have a user database of about 1 million users and we noticed that resetting your password is rather slow. It turns out that there isn't any index on neither the password reset nor the verification token.

## The solution
The simplest solution would be to add an index to those fields and call it a day. But we can do even better. We know that those tokens are unique. Because all throughout the Sylius application those fields are queried with the `findOneBy` which crashes on result sets bigger than one. So adding a unique constraint should not break backwards compatibility (except for already broken entries).

## What still needs to be done
I still need to generate a migration that actually adds those indices to the database. I will do that when the base idea of the PR is accepted.

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\CoreBundle\Migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php]: Please remove this message

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+final class Version20231103004216 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return 'Making the uniqueness of the password_reset_token and email_verification_token a database constraint';
+    }
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php]: Please remove this message

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        $this->addSql('CREATE UNIQUE INDEX UNIQ_88D5CC4D6B7BA4B6 ON sylius_admin_user (password_reset_token)');
+        $this->addSql('CREATE UNIQUE INDEX UNIQ_88D5CC4DC4995C67 ON sylius_admin_user (email_verification_token)');
+        $this->addSql('CREATE UNIQUE INDEX UNIQ_7C2B74806B7BA4B6 ON sylius_shop_user (password_reset_token)');
+        $this->addSql('CREATE UNIQUE INDEX UNIQ_7C2B7480C4995C67 ON sylius_shop_user (email_verification_token)');
+    }
+
+    public function down(Schema $schema): void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php]: Please remove this message

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20231103004216 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return 'Making the uniqueness of the password_reset_token and email_verification_token a database constraint';
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php]: Looking at recent migrations (As well as what I usually do), description should be similar in style to commit message.

For example,
```suggestion
        return 'Add unique indices for password_reset_token, email_verification_token fields to sylius_shop_user and sylius_admin_user tables';
```

File: src/Sylius/Bundle/UserBundle/Resources/config/doctrine/model/User.orm.xml [hunk 5]
```diff
@@ -25,9 +25,9 @@
         <field name="encoderName" column="encoder_name" type="string" nullable="true" />
 
         <field name="lastLogin" column="last_login" type="datetime" nullable="true" />
-        <field name="passwordResetToken" column="password_reset_token" type="string" nullable="true" />
+        <field name="passwordResetToken" column="password_reset_token" type="string" nullable="true" unique="true" />
```
Comment (diimpp) [inline src/Sylius/Bundle/UserBundle/Resources/config/doctrine/model/User.orm.xml]: While there are no standards, it's more common for sylius mappings to specify `nullable` last. Here and in `emailVerificationToken`

image

**Review by diimpp — COMMENTED** [review]
Apparently unique index on `nullable` field is a sort of grey zone, but which works for mysql. Is it confirmed to work with postgresql as well?

https://dba-presents.com/databases/mysql/274-mysql-unique-index-with-null-values

File: src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
+<?php
+/*
```
Comment (diimpp) [inline src/Sylius/Bundle/CoreBundle/Migrations/Version20231103004216.php:3]: Missing empty line between lines 1 and 2.

**Review by diimpp — APPROVED** [review]

File: src/Sylius/Bundle/UserBundle/Resources/config/doctrine/model/User.orm.xml [hunk 7]
```diff
@@ -25,9 +25,9 @@
         <field name="encoderName" column="encoder_name" type="string" nullable="true" />
 
         <field name="lastLogin" column="last_login" type="datetime" nullable="true" />
-        <field name="passwordResetToken" column="password_reset_token" type="string" nullable="true" />
+        <field name="passwordResetToken" column="password_reset_token" type="string" unique="true" nullable="true" />
         <field name="passwordRequestedAt" column="password_requested_at" type="datetime" nullable="true" />
-        <field name="emailVerificationToken" column="email_verification_token" type="string" nullable="true" />
+        <field name="emailVerificationToken" column="email_verification_token" type="string"  nique="true" nullable="true" />
```
Comment (diimpp) [inline src/Sylius/Bundle/UserBundle/Resources/config/doctrine/model/User.orm.xml]:
```suggestion
        <field name="emailVerificationToken" column="email_verification_token" type="string" unique="true" nullable="true" />
```

**Review by diimpp — CHANGES_REQUESTED** [review]

File: src/Sylius/Bundle/UserBundle/Resources/config/doctrine/model/User.orm.xml (hunk 7 above)
Comment (mamazu) [reply src/Sylius/Bundle/UserBundle/Resources/config/doctrine/model/User.orm.xml]: Fixed it manually. It's very odd that the tests don't fail with that.

Comment (diimpp) [reply src/Sylius/Bundle/UserBundle/Resources/config/doctrine/model/User.orm.xml]: > Fixed it manually. It's very odd that the tests don't fail with that.

Yep, no idea how it passed.

**Review by diimpp — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Comment by NoResponseMate** [comment]
Thank you, @mamazu!

**Pull request merged**